File: backend/src/stack/loader.ts
import { readFileSync, readdirSync, existsSync } from "fs";
import { join } from "path";
import type { StackManifest } from "../types";
import { isStackManifest, validateStackManifest } from "./schema";

/** Built-in demo stack (always available even if YAML is unavailable at runtime). */
export const DEMO_STACK_MANIFEST: StackManifest = {
  apiVersion: "kip.ai/v1",
  kind: "StackManifest",
  metadata: { name: "demo-prod-gpu-west-2" },
  spec: {
    displayName: "prod-gpu-west-2",
    providers: {
      cluster: { type: "demo" },
      gpu: { type: "demo" },
      observability: { type: "demo" },
      cost: { type: "demo" },
      policy: { type: "demo" },
      runtime: { type: "demo" },
      remediation: { type: "demo" },
    },
  },
};

/**
 * Minimal YAML subset parser for Stack Manifest files.
 * Supports the flat structure used by KIP stacks (no anchors/multiline).
 */
export function parseSimpleYaml(source: string): unknown {
  const lines = source
    .split(/\r?\n/)
    .map((l) => l.replace(/#.*$/, ""))
    .filter((l) => l.trim().length > 0);

  type Node = Record<string, unknown>;
  const root: Node = {};
  const stack: Array<{ indent: number; node: Node }> = [{ indent: -1, node: root }];

  for (const raw of lines) {
    const indent = raw.match(/^\s*/)?.[0].length ?? 0;
    const trimmed = raw.trim();
    const match = trimmed.match(/^([A-Za-z0-9_-]+):\s*(.*)$/);
    if (!match) continue;

    const [, key, valueRaw] = match;
    while (stack.length > 1 && indent <= stack[stack.length - 1].indent) {
      stack.pop();
    }
    const parent = stack[stack.length - 1].node;

    if (valueRaw === "") {
      const child: Node = {};
      parent[key] = child;
      stack.push({ indent, node: child });
    } else {
      const value = coerceScalar(valueRaw);
      parent[key] = value;
    }
  }

  return root;
}

function coerceScalar(raw: string): unknown {
  const v = raw.trim();
  if ((v.startsWith('"') && v.endsWith('"')) || (v.startsWith("'") && v.endsWith("'"))) {
    return v.slice(1, -1);
  }
  if (v === "true") return true;
  if (v === "false") return false;
  if (/^-?\d+(\.\d+)?$/.test(v)) return Number(v);
  return v;
}

export function loadStackManifestFromYaml(source: string): StackManifest {
  const parsed = parseSimpleYaml(source);
  if (!isStackManifest(parsed)) {
    throw new Error("Invalid StackManifest YAML");
  }
  const errors = validateStackManifest(parsed);
  if (errors.length) {
    throw new Error(`Invalid StackManifest: ${errors.join("; ")}`);
  }
  return parsed;
}

function stacksDir(): string {
  return join(process.cwd(), "backend", "stacks");
}

export function loadAllStackManifests(): StackManifest[] {
  const byName = new Map<string, StackManifest>();
  byName.set(DEMO_STACK_MANIFEST.metadata.name, DEMO_STACK_MANIFEST);

  const dir = stacksDir();
  if (!existsSync(dir)) {
    return Array.from(byName.values());
  }

  for (const file of readdirSync(dir)) {
    if (!file.endsWith(".yaml") && !file.endsWith(".yml")) continue;
    try {
      const source = readFileSync(join(dir, file), "utf8");
      const manifest = loadStackManifestFromYaml(source);
      byName.set(manifest.metadata.name, manifest);
    } catch {
      // Skip invalid files; demo stack remains available.
    }
  }

  return Array.from(byName.values());
}

export function getStackManifest(stackId: string): StackManifest | undefined {
  return loadAllStackManifests().find((m) => m.metadata.name === stackId);
}

File: backend/kip/stack/loader.py
from __future__ import annotations

from pathlib import Path
from typing import Any

import yaml

from kip.types import ProviderRef, StackManifest, StackMetadata, StackSpec, CAPABILITY_KINDS

DEMO_STACK = StackManifest(
    metadata=StackMetadata(name="demo-prod-gpu-west-2"),
    spec=StackSpec(
        displayName="prod-gpu-west-2",
        providers={
            "cluster": ProviderRef(type="demo"),
            "gpu": ProviderRef(type="demo"),
            "observability": ProviderRef(type="demo"),
            "cost": ProviderRef(type="demo"),
            "policy": ProviderRef(type="demo"),
            "runtime": ProviderRef(type="demo"),
            "remediation": ProviderRef(type="demo"),
        },
    ),
)


def _stacks_dir() -> Path:
    return Path(__file__).resolve().parent.parent.parent / "stacks"


def _from_dict(data: dict[str, Any]) -> StackManifest:
    providers_raw = data["spec"]["providers"]
    providers = {
        kind: ProviderRef(**providers_raw[kind])
        if isinstance(providers_raw[kind], dict)
        else ProviderRef(type=str(providers_raw[kind]))
        for kind in CAPABILITY_KINDS
        if kind in providers_raw
    }
    return StackManifest(
        apiVersion=data.get("apiVersion", "kip.ai/v1"),
        kind=data.get("kind", "StackManifest"),
        metadata=StackMetadata(name=data["metadata"]["name"]),
        spec=StackSpec(
            displayName=data["spec"]["displayName"],
            providers=providers,  # type: ignore[arg-type]
        ),
    )


def load_all_stack_manifests() -> list[StackManifest]:
    by_name: dict[str, StackManifest] = {DEMO_STACK.metadata.name: DEMO_STACK}
    directory = _stacks_dir()
    if not directory.exists():
        return list(by_name.values())

    for path in directory.glob("*.y*ml"):
        try:
            data = yaml.safe_load(path.read_text())
            if not isinstance(data, dict):
                continue
            manifest = _from_dict(data)
            by_name[manifest.metadata.name] = manifest
        except Exception:
            continue
    return list(by_name.values())


def get_stack_manifest(stack_id: str) -> StackManifest | None:
    return next(
        (m for m in load_all_stack_manifests() if m.metadata.name == stack_id),
        None,
    )
